configserver/src/main/java/com/yahoo/vespa/config/server/tenant/TenantRepository.java

```diff
@@ -612,4 +615,36 @@ public static Path getBarriersPath() {
 
     public Curator getCurator() { return curator; }
 
+    /**
+     * Single-threaded executor (to make sure that getting data for a ZooKeeper watcher event is done serially) that
+     * delegates to another executor (that uses a cached thread pool)
+     */
+    private static class ZkWatcherExecutorService extends ThreadPoolExecutor {
+
+        private final ExecutorService executorService =
+                Executors.newCachedThreadPool(new DaemonThreadFactory("zk-session-watcher-"));
+
+        public ZkWatcherExecutorService() {
+            super(1, 1, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>());
+        }
+
+        @Override
+        public void execute(Runnable command) {
+            super.execute(() -> executorService.execute(command));
+        }
+
+        @Override
+        public void shutdown() {
+            super.shutdown();
+            executorService.shutdown();
+        }
+
+        @Override
+        public List<Runnable> shutdownNow() {
+            super.shutdownNow();
+            return executorService.shutdownNow();
+        }
```

Comment:
Could also delegate `awaitTermination` to both parties.

Reply:
Yup, will add